[SourceKit] Add a dedicated request to retrieve the diagnostics of a file

ahoppen opened the pull request:
Currently, SourceKit always implicitly sends diagnostics to the client after every edit. This doesn’t match our new cancellation story because diagnostics retrieval is no request and thus can’t be cancelled.

Instead, diagnostics retrieval should be a standalone request, which returns a cancellation token like every other request and which can thus also be cancelled as such.

The indented transition path here is to change all open and edit requests to be syntactic only. When diagnostics are needed, the new diagnostics request should be used.

rdar://83391522

ahoppen committed:
[SourceKit] Add a dedicated request to retrieve the diagnostics of a file

Currently, SourceKit always implicitly sends diagnostics to the client after every edit. This doesn’t match our new cancellation story because diagnostics retrieval is no request and thus can’t be cancelled.

Instead, diagnostics retrieval should be a standalone request, which returns a cancellation token like every other request and which can thus also be cancelled as such.

The indented transition path here is to change all open and edit requests to be syntactic only. When diagnostics are needed, the new diagnostics request should be used.

rdar://83391522

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test

benlangmuir commented on tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:
```diff
@@ -10,12 +10,13 @@
 #include "SourceKit/Support/UIdent.h"
+#include "SwiftASTManager.h"
+#include "SwiftEditorDiagConsumer.h"
+#include "SwiftLangSupport.h"
```
Nit: We usually sort implementation headers outside include directories to the top of the list, like it was before.

ahoppen replied:
I just accepted what `git-clang-format` offered. Do you think it’s better to keep the old sorting or give in to `clang-format`?

benlangmuir replied:
Personally I would keep the old sorting, but I'm not worried about it either way if you disagree

benlangmuir commented on tools/SourceKit/tools/sourcekitd-test/sourcekitd-test.cpp:
```diff
@@ -1247,7 +1260,8 @@ static bool handleResponse(sourcekitd_response_t Resp, const TestOptions &Opts,
   } else if (Opts.PrintRawResponse) {
-    sourcekitd_response_description_dump_filedesc(Resp, STDOUT_FILENO);
+    llvm::outs().flush();
+    printRawResponse(Resp);
```
Unnecessary flush since you're doing it in `printRawResponse` itself?

benlangmuir reviewed (approved):
Two minor comments, but otherwise LGTM.

ahoppen committed:
[SourceKit] Flush llvm::outs() before printing to STDOUT_FILENO in soucekitd-test

Otherwise, we get non-deterministic ordering of results printed through `llvm::outs()` and `STDOUT_FILENO`.

ahoppen commented:
@swift-ci Please smoke test
